src/vsr/journal.zig

```diff
@@ -2096,21 +2110,19 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
             while (i < headers_per_sector) : (i += 1) {
                 const slot = Slot{ .index = sector_slot.index + i };
 
-                if (journal.faulty.bit(slot)) {
-                    // Redundant faulty headers are deliberately written as invalid.
-                    // This ensures that faulty headers are still faulty when they are read back
-                    // from disk during recovery. This prevents faulty entries from changing to
-                    // reserved (and clean) after a crash and restart (e.g. accidentally converting
-                    // a case `@D` to a `@I` after a restart).
-                    sector_headers[i] = Header.Prepare.reserved(replica.cluster, i);
-                    sector_headers[i].checksum = 0; // Invalidate the checksum.
-                    assert(!sector_headers[i].valid_checksum());
+                if (journal.headers_redundant[slot.index].operation == .reserved and
+                    journal.headers_redundant[slot.index].checksum == 0)
+                {
+                    // Deliberately write an invalid header until the corresponding prepare is
+                    // repaired. (See read_prepare_with_op_and_checksum_callback()).
+                    assert(journal.faulty.bit(slot));
                 } else {
-                    // Write headers from `headers_redundant` instead of `headers` — we need to
-                    // avoid writing (leaking) a redundant header before its corresponding prepare
-                    // is on disk.
-                    sector_headers[i] = journal.headers_redundant[slot.index];
+                    maybe(journal.faulty.bit(slot));
                 }
+
+                // Write headers from `headers_redundant` instead of `headers` — we need to avoid
+                // writing (leaking) a redundant header before its corresponding prepare is on disk.
+                sector_headers[i] = journal.headers_redundant[slot.index];
```

Comment:
Consider doing a bulk copy_disjoint here, rather than doing it one header at a time, to make it clear that what we write is `headers_redundant` as is (but copied to avoid races).

Unrelated, but the `headers_iops` name is a bit confusing, though not sure what would be better. `write_headers_sector`?